docs(contribution): add experimental features policy, clarify ai contributions, restructure (#2943)
Redocly/redocly-cli: CONTRIBUTING.md

**Maintainers (when merging):** remove redundant intermediate commit descriptions and keep the main commit description (plus co-authors if needed).
If several commit descriptions should be preserved, rebase instead of squashing.

### Experimental features

We accept contributions for experimental features, but they should be clearly marked as such in the documentation and changeset.
Checklist for experimental features:

1. Clean code isolation (preferably one entrypoint to the existing code).
1. Reuse of existing approaches, concepts, patterns, types, functions, and constants (such as walker, `NodeType`, `detectSpec`, `isRef`, etc.).

> 1. Avoid unnecessary concept duplication.

Reviewer: ```suggestion
1. No unnecessary concept duplication.
```

1. Clean, concise docs.

> 1. Avoid excessive unit tests (especially testing internal implementation details), prefer e2e; you can request the coverage threshold adjustment if needed.

Reviewer: ```suggestion
1. No excessive unit tests (especially testing internal implementation details), prefer e2e.
    You can request the coverage threshold adjustment if needed.
```

| 1         | Command execution error   |
| 2         | Config resolution failure |

### Local source code usage

> There are two options for testing local changes in other local applications: NPM linking and local packing and installing from the `redocly-cli.tgz` file.

Reviewer: ```suggestion
There are two options for testing local changes in other local applications:
- NPM linking
- local packing and installing from the `redocly-cli.tgz` file
```

#### NPM linking

To test the local source code of the packages in other local applications, you can use npm linking.

> See the [docs](https://docs.npmjs.com/cli/v9/commands/npm-link).

Reviewer: ```suggestion
To learn how to use npm linking, see the [offical docs](https://docs.npmjs.com/cli/v9/commands/npm-link).
```

#### Local packing and installing

To test local changes as a package, you can use the following steps:

1. Optionally, change the version of the packages.

1. Run `npm run pack:prepare` in the repository's root.
   This generates **redocly-cli.tgz**, **respect-core.tgz**, and **openapi-core.tgz** files.

1. Copy those **.tgz** files to a destination folder and then run `npm install redocly-cli.tgz` there to install Redocly CLI.

>    To install `openapi-core` do the same but with **openapi-core.tgz** file.

Reviewer: ```suggestion
1. (Optional) Change the version of the packages.

1. Run `npm run pack:prepare` in the repository's root.
   This command generates **redocly-cli.tgz**, **respect-core.tgz**, and **openapi-core.tgz** files.

1. Copy these **.tgz** files to a destination folder and then run `npm install redocly-cli.tgz` there to install Redocly CLI.
   To install `openapi-core`: repeat this step, but with **openapi-core.tgz** file.
```

  [Learn more about mTLS usage in Respect](https://redocly.com/docs/respect/guides/mtls-cli#use-mtls-with-respect-in-redocly-cli).
- File upload is working for both `multipart/form-data` and `application/octet-stream`.

## Contribute documentation

Additions and updates to our documentation are very welcome.
You can find the documentation in the `docs/` folder, and this is published to https://redocly.com/docs/cli/ as part of our main website.

To preview your documentation changes locally:

1. Make sure `redocly` is already installed on your local computer.
   See [installation](https://redocly.com/docs/cli/installation/).

1. Run this command from the `docs/` folder:

   ```bash
   redocly preview
   ```

By default, you can access the docs preview at http://localhost:4000 or http://127.0.0.1:4000.

> Please note that currently the custom markdoc tags used in the main website are not available in the local preview version, and links that point to the wider website do show as errors when using a local platform.

> > The pull request workflows generate a full preview, so rest assured that you are able to check everything is in good shape before we review and merge your changes.

Reviewer: ```suggestion
> Currently, the custom Markdoc tags used in the main website are not available in the local preview version, and links that point to the wider website do show as errors when using a local platform.
> Pull request workflows generate full previews: you will be able to check if everything is in good shape before we review and merge your changes.
```

We are proud of our docs.
When you open a pull request, we lint the prose using [Vale](https://vale.sh/).
You can also install this tool locally and run it from the root of the project with:

```bash
vale README.md docs/ .changeset
```

The configuration is in `.vale.ini` in the root of the project.

### Markdown linting

We use [Markdownlint](https://github.com/DavidAnson/markdownlint-cli2) to check that the Markdown in our docs is well formatted (config: `.markdownlint.yaml`).
The checks run on pull requests; locally, from the repository root:

```bash
npx markdownlint-cli2 "docs/**/*.md"
```

> Note that formatter also runs and reformats Markdown files.
> Use `npm run format` from the root of the project.

> Please use semantic line breaks in markdown files (preferably splitting sentences into separate lines).

### Markdown link checking

We use [`mlc`](https://github.com/becheran/mlc) to check the links in the `docs/` folder.
This tool runs automatically on every pull request, but you can also run it locally if you want to.
Visit the project homepage to find the installation instructions for your platform, and then run the command like this:

```bash
mlc docs/
```

> It only checks links within the local docs (it can't check links to other docs sections that are present when we publish all products under https://redocly.com/docs), and doesn't currently check anchors, so take care when renaming pages or titles.

Reviewer: ```suggestion
The tool only checks links within the local docs (it can't check links to other docs sections that are present when we publish all products under https://redocly.com/docs), and doesn't currently check anchors.
Take care when renaming pages or titles.
```
